File: Zero-to-hero/Python/875_koko_eating_bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        if len(piles) ==  h:
            return max(piles)
        def curTime(k, piles):
            cur = 0
            for pile in piles:
                if pile <= k:
                    cur += 1
                else:
                    cur += math.ceil(pile/k)
            return cur
        left = 1
        right = max(piles)
        while left < right:
            mid = left + (right - left) // 2
            time = curTime(mid, piles)
            if time > h:
                left = mid + 1
            else:
                right = mid
        return left

File: Zero-to-hero/Java/875_koko_eating_bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int left =1;
        int maxV = 0;
        for (int i=0; i < piles.length; i++){
            if (maxV < piles[i]){
                maxV = piles[i];
            }
        }
        int right = maxV;
        while (left < right){
            int mid = left + (right - left)/2;
            int time = curTime(piles, mid);
            if (time > h){
                left = mid + 1;
            }else{
                right = mid;
            }
        }
        return left;

    }
    public int curTime(int [] piles, int k){
        int cur = 0;
        for (int pile: piles){
            if (k >= pile){
                cur += 1;
            }else{
                cur += (k + pile -1) / k;
            }
        }
        return cur;
    }
}